utils_nlp/models/transformers/question_answering.py

```diff
@@ -0,0 +1,1647 @@
+# coding=utf-8
+# Copyright 2018 The Google AI Language Team Authors and The HuggingFace Inc. team.
+# Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Modifications copyright © Microsoft Corporation
+
+
+import os
+import logging
+from tqdm import tqdm
+import collections
+import json
+import math
+import jsonlines
+
+import torch
+from torch.utils.data import TensorDataset, SequentialSampler, DataLoader
+
+from transformers.tokenization_bert import BasicTokenizer, whitespace_tokenize
+from transformers import XLNetTokenizer
+
+from transformers.modeling_bert import (
+    BertConfig,
+    BERT_PRETRAINED_MODEL_ARCHIVE_MAP,
+    BertForQuestionAnswering,
+)
+
+from transformers.modeling_xlnet import (
+    XLNetConfig,
+    XLNET_PRETRAINED_MODEL_ARCHIVE_MAP,
+    XLNetForQuestionAnswering,
+)
+
+from transformers.modeling_distilbert import (
+    DistilBertConfig,
+    DISTILBERT_PRETRAINED_MODEL_ARCHIVE_MAP,
+    DistilBertForQuestionAnswering,
+)
+
+from utils_nlp.models.transformers.common import (
+    MAX_SEQ_LEN,
+    TOKENIZER_CLASS,
+    Transformer,
+    get_device,
+)
+
+MODEL_CLASS = {}
+MODEL_CLASS.update({k: BertForQuestionAnswering for k in BERT_PRETRAINED_MODEL_ARCHIVE_MAP})
+MODEL_CLASS.update({k: XLNetForQuestionAnswering for k in XLNET_PRETRAINED_MODEL_ARCHIVE_MAP})
+MODEL_CLASS.update(
+    {k: DistilBertForQuestionAnswering for k in DISTILBERT_PRETRAINED_MODEL_ARCHIVE_MAP}
+)
+
+CONFIG_CLASS = {}
+CONFIG_CLASS.update({k: BertConfig for k in BERT_PRETRAINED_MODEL_ARCHIVE_MAP})
+CONFIG_CLASS.update({k: XLNetConfig for k in XLNET_PRETRAINED_MODEL_ARCHIVE_MAP})
+CONFIG_CLASS.update({k: DistilBertConfig for k in DISTILBERT_PRETRAINED_MODEL_ARCHIVE_MAP})
+
+# cached files during preprocessing
+# these are used in postprocessing to generate the final answer texts
+CACHED_EXAMPLES_TRAIN_FILE = "cached_examples_train.jsonl"
+CACHED_FEATURES_TRAIN_FILE = "cached_features_train.jsonl"
+
+CACHED_EXAMPLES_TEST_FILE = "cached_examples_test.jsonl"

```

Comment:
can't the examples stay in memory? are you persisting them because of memory limitation?

Reply:
These are used when you call postprocess_bert/xlnet_answer. 
I used to return these from the preprocess method, the caching file idea came up when we previously designed preprocess to be an iterator. That's also why I used jsonl instead of json, so that I can keep adding new data to the file every time a new batch is yield. 
Now that you mentioned it, it does help saving some memory that can be used for training.